ITE: drivers/i2c: I2C setting frequency adjustment

GTLin08 opened the pull request:
In order to pass the SI (Signal Integrity) test, the timing registers of  I2C 0/1/2 port
can be used to adjust tHD; DAT and tSU;DAT to pass it, and I2C 3/4/5 port can add
a config of CONFIG_I2C_ADJUST_SCL_LOW_PRESCALE to adjust the low period of
SCL clock to pass tLOW test.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
ITE: drivers/i2c: add the config to adjust SCL low period prescale

This config can adjust prescale of I2C3/4/5 port to set I2C speed
for SCL low period.
If CONFIG_I2C_ADJUST_SCL_LOW_PRESCALE=1 will increase the low
period of the SCL clock but reduce the frequency.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

teburd commented:
I would think signal integrity issues would be board specific, having this Kconfig seems reasonable to me to allow for tweaking

teburd commented on drivers/i2c/Kconfig.it8xxx2:
```diff
@@ -8,3 +8,16 @@ config I2C_ITE_IT8XXX2
 	  This driver supports repeated start.
+
+config I2C_ADJUST_SCL_LOW_PRESCALE
+	int "Adjust SCL low period prescale"
```
It would be good to have a range attribute for this, as I don't think any integer is useful here is it? At the very least looks to be capped 0 to 255

GTLin08 replied:
Done.

teburd commented on drivers/i2c/i2c_ite_it8xxx2.c:
```diff
@@ -242,10 +242,10 @@ static void i2c_standard_port_timing_regs_400khz(uint8_t port)
 	IT83XX_SMB_SCLKTS(port) = 0;
 	/* Suggested setting of timing registers of 400kHz. */
-	IT83XX_SMB_4P7USL = 0x6;
+	IT83XX_SMB_4P7USL = 0x3;
```
Are these documented somewhere? A quick search on ITE's site I didn't see any docs on what these mean or what values should be here. I'm curious why the suggest values are changing here, if its to correcting for a board specific problem this isn't the place to do that.

GTLin08 replied:
<img width="559" alt="timing_reg" src="https://user-images.githubusercontent.com/73827039/154918790-a4c08e5f-a0a7-4b33-976c-2ad020d06b4a.png">
<img width="620" alt="timing_reg2" src="https://user-images.githubusercontent.com/73827039/154918799-d3ff7013-9db8-4c50-ae93-4afa0d4e61a5.png">

As shown above, we can control the TSU_STA and THD_DAT simultaneously by changing the value of the register 4P7USL. Likewise, we can control the tsu_dat by changing the value of the register 250NS as well.

The lower the value of 4P7USL is set, the shorter the THD_DAT we will have. Likewise, the higher we configure the 250NS, the longer the tsu_dat will be. Therefore, we can configure the 4P7USL and the 250NS properly so that we can keep the Tlow unchanged.

teburd replied:
Thank you for the great explanation, that makes perfect sense as to what these values are doing now.

teburd reviewed (commented):
It makes some sense as a way to tweak the frequency somewhat through Kconfig, though the other PR might be worth looking at closer for arbitrary i2c frequencies through device tree. It's not like i2c *must* run at some frequency, there's usually limits in part datasheets for timings not exact "must be 100kHz"

GTLin08 commented:
Hi @keith-zephyr , @brockus-zephyr 
This PR needs your review, thanks!

gmarull commented on drivers/i2c/Kconfig.it8xxx2:
```diff
@@ -8,3 +8,17 @@ config I2C_ITE_IT8XXX2
 	  Supported Speeds: 100kHz, 400kHz and 1MHz.
 	  This driver supports repeated start.
+
+config I2C_ADJUST_SCL_LOW_PRESCALE
```
1. Why can't this be in Devicetree?
2. This option has no dependency set on IT8XXX2. By looking at it, it's not clear it's an option specific to IT8XXX2.

GTLin08 replied:
1. Due to the fact that the logic design of I2C channel 0, 1, 2 is different from the design of channel 3, 4, 5, they are two different logic blocks in the silicon. Hence, the config here only works for adjusting the frequency of I2C3, I2C4, and I2C5. If we add the properties into the devicetree, I’m afraid that users could probably misunderstand that the frequency of I2C0, 1, 2 could be adjusted through it but actually it’d not work and I don’t know if it’d cause side effects. It’s also mentioned in the line 17 of the drivers/i2c/Kconfig.it8xxx2, we made some descriptions on this.
2. Done.

gmarull commented on drivers/i2c/Kconfig.it8xxx2:
```diff
@@ -8,3 +8,18 @@ config I2C_ITE_IT8XXX2
+	default 0
+	range 0 255
+	help
+	  This config can adjust prescale of I2C3/4/5 port to set
+	  I2C speed for SCL low period.
+	  If CONFIG_I2C_ADJUST_SCL_LOW_PRESCALE=1 will increase
+	  the low period of the SCL clock but reduce the frequency.
+	  The formula is as follows.
+	  1 SCL cycle =
+	  2 x (psr + I2C_ADJUST_SCL_LOW_PRESCALE + 2) x
+	  SMBus clock cycle
```
```suggestion

	  This option is used to configure the I2C speed prescaler for the SCL low period. It only applies to I2C3/4/5 ports.
	  When set to >= 1, it will increase the low period of the SCL clock and so reduce the signal frequency.
	  The resulting SCL cycle time is given by the following formula:
	  SCL cycle =
	  2 * (psr + I2C_ADJUST_SCL_LOW_PRESCALE + 2) *
	  SMBus clock cycle
```

(may need to be adjusted, just trying to make it more clear)

GTLin08 replied:
Thanks for your advice. Done.

GTLin08 committed:
ITE: drivers/i2c: adjust timing registers to pass SI test

Timing registers of I2C 0/1/2 can be adjusted to pass SI test.

We can control the tSU;STA and tHD;DAT simultaneously by changing
the value of the register IT83XX_SMB_4P7USL, and we can control
the tSU;DAT by changing the value of the register IT83XX_SMB_250NS
as well.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 commented on drivers/i2c/Kconfig.it8xxx2:
```diff
@@ -8,3 +8,18 @@ config I2C_ITE_IT8XXX2
 	  Supported Speeds: 100kHz, 400kHz and 1MHz.
 	  This driver supports repeated start.
+
+config I2C_ADJUST_SCL_LOW_PRESCALE
```
Refer to other Kconfig, and add the prefix of ITE. Thanks!

mbolivar-nordic commented:
> Due to the fact that the logic design of I2C channel 0, 1, 2 is different from the design of channel 3, 4, 5, they are two different logic blocks in the silicon.

Then they should have different devicetree compatibles.

mbolivar-nordic reviewed (changes requested):
I'm not convinced about this. If there are different IP blocks, then there should be different DT compatibles. Then you can move this configuration to DT where it really seems like it belongs. What am I missing?

henrikbrixandersen commented:
> I'm not convinced about this. If there are different IP blocks, then there should be different DT compatibles. Then you can move this configuration to DT where it really seems like it belongs.

I tend to agree that this seems like a better solution.

mbolivar-nordic commented:
@GTLin08 please update the PR accordingly, then. Notice that there is precedent for this in e.g. the following STM32 I2C controller bindings:

- https://docs.zephyrproject.org/latest/reference/devicetree/bindings/i2c/st%2Cstm32-i2c-v1.html#dtbinding-st-stm32-i2c-v1
- https://docs.zephyrproject.org/latest/reference/devicetree/bindings/i2c/st%2Cstm32-i2c-v2.html#dtbinding-st-stm32-i2c-v2

You can take a similar approach of using `-v1` and `-v2` endings if your hardware is along similar lines of different versions of the IP block. If it's not, feel free to pick more descriptive compatible values.

GTLin08 commented:
Thanks for your review, I have created another PR #43357 for I2C different compatibles, so close this.
